File: HelloWorld/task3/CsDfsAndBfs/CsDfsAndBfs/Services/Algorithms.cs
using System.Collections.Generic;
using CsDfsAndBfs.Entities;
using CsDfsAndBfs.Tools;

namespace CsDfsAndBfs.Services
{
    public class Algorithms
    {
		static public Graph<ReturnType> Dfs<Type, ReturnType>(
			Graph<Type> graph,
			int vertex,
			IUnaryOperation<Type, ReturnType> operation)
		{
			var resultGraph = new Graph<ReturnType>(graph.Size());
			resultGraph.ConstructByGraph(graph);
			var graphCopy = new Graph<bool>(graph.Size());
			graphCopy.ConstructByGraph(graph);
			for (int i = 0; i < graphCopy.Size(); ++i)
			{
				graphCopy.ChangeVertex(i, false);
			}
			Dfs(graph, vertex, operation, resultGraph, graphCopy);
			return resultGraph;
		}

		static public Graph<ReturnType> Bfs<Type, ReturnType>(
				Graph<Type> graph,
				int vertex,
				IUnaryOperation<Type, ReturnType> operation)
		{
			var resultGraph = new Graph<ReturnType>(graph.Size());
			resultGraph.ConstructByGraph(graph);
			var graphCopy = new Graph<bool>(graph.Size());
			graphCopy.ConstructByGraph(graph);
			for (int i = 0; i < graphCopy.Size(); ++i)
			{
				graphCopy.ChangeVertex(i, false);
			}
			Bfs(graph, vertex, operation, resultGraph, graphCopy);
			return resultGraph;
		}

		static private void Dfs<Type, ReturnType>(
				Graph<Type> graph,
				int vertex,
				IUnaryOperation<Type, ReturnType> operation,
				Graph<ReturnType> resultGraph,
				Graph<bool> graphCopy)
		{
			graphCopy.ChangeVertex(vertex, true);
			resultGraph.ChangeVertex(vertex, operation.Apply(graph.GetVertex(vertex)));
			var neighbours = graphCopy.GetVertexNeighbours(vertex);
			foreach (int neighbour in neighbours)
			{
				if (!graphCopy.GetVertex(neighbour))
				{
					Dfs(graph, neighbour, operation, resultGraph, graphCopy);
				}
			}
		}

		static private void Bfs<Type, ReturnType>(
				Graph<Type> graph,
				int vertex,
				IUnaryOperation<Type, ReturnType> operation,
				Graph<ReturnType> resultGraph,
				Graph<bool> graphCopy)
		{
			var vertexesQueue = new Queue<int>();
			vertexesQueue.Enqueue(vertex);
			while (vertexesQueue.Count != 0)
			{
				int vertexNow = vertexesQueue.Dequeue();
				graphCopy.ChangeVertex(vertexNow, true);
				resultGraph.ChangeVertex(vertexNow, operation.Apply(graph.GetVertex(vertexNow)));
				var neighbours = graphCopy.GetVertexNeighbours(vertexNow);
				foreach (int neighbour in neighbours)
				{
					if (!graphCopy.GetVertex(neighbour))
					{
						vertexesQueue.Enqueue(neighbour);
					}
				}
			}
		}
	}
}

File: HelloWorld/task3/JavaDfsAndBfs/graphs/src/main/java/dfs_and_bfs1/graphs/Algorithms.java
package dfs_and_bfs1.graphs;
import java.util.*;

public class Algorithms {
	static public<Type, ReturnType> Graph<ReturnType> dfs(
			Graph<Type> graph,
			int vertex,
			IUnaryOperation<Type, ReturnType> operation)
	{
		Graph<ReturnType> resultGraph = new Graph<ReturnType>(graph);
		Graph<Boolean> graphCopy = new Graph<Boolean>(graph);
		for (int i = 0; i < graphCopy.size(); ++i)
		{
			graphCopy.changeVertex(i, false);
		}
	    dfs(graph, vertex, operation, resultGraph, graphCopy);
	    return resultGraph;
	}
	
	static public<Type, ReturnType> Graph<ReturnType> bfs(
			Graph<Type> graph,
			int vertex,
			IUnaryOperation<Type, ReturnType> operation)
	{
		Graph<ReturnType> resultGraph = new Graph<ReturnType>(graph);
		Graph<Boolean> graphCopy = new Graph<Boolean>(graph);
		for (int i = 0; i < graphCopy.size(); ++i)
		{
			graphCopy.changeVertex(i, false);
		}
		bfs(graph, vertex, operation, resultGraph, graphCopy);
		return resultGraph;
	}
	
	static private<Type, ReturnType> void dfs(
			Graph<Type> graph,
			int vertex,
			IUnaryOperation<Type, ReturnType> operation,
			Graph<ReturnType> resultGraph,
			Graph<Boolean> graphCopy)
	{
		graphCopy.changeVertex(vertex, true);
		resultGraph.changeVertex(vertex, operation.apply(graph.getVertex(vertex)));
		Vector<Integer> neighbours = graphCopy.getVertexNeighbours(vertex);
		for (int neighbour : neighbours)
		{
			if (!graphCopy.getVertex(neighbour))
			{
				dfs(graph, neighbour, operation, resultGraph, graphCopy);
			}
		}
	}
	
	static private<Type, ReturnType> void bfs(
			Graph<Type> graph,
			int vertex,
			IUnaryOperation<Type, ReturnType> operation,
			Graph<ReturnType> resultGraph,
			Graph<Boolean> graphCopy)
	{
		Queue<Integer> vertexesQueue = new ArrayDeque<Integer>(); 
		vertexesQueue.add(vertex);
		while (!vertexesQueue.isEmpty())
		{
			int vertexNow = vertexesQueue.poll();
			graphCopy.changeVertex(vertexNow, true);
			resultGraph.changeVertex(vertexNow, operation.apply(graph.getVertex(vertexNow)));
			Vector<Integer> neighbours = graphCopy.getVertexNeighbours(vertexNow);
			for (int neighbour : neighbours)
			{
				if (!graphCopy.getVertex(neighbour))
				{
					vertexesQueue.add(neighbour);
				}
			}
		}
	}
}
